test/upgrade/2.2.0-2.2.1.js

```diff
@@ -0,0 +1,200 @@
+  let accountHandler;
+  let snapshot;
+  let protocolDiamondAddress, mockContracts;
+
+  // reference protocol state
+  let accountContractState;
+  let preUpgradeEntities;
+
+  before(async function () {
+    try {
+      // Make accounts available
+      [deployer, rando] = await ethers.getSigners();
+
+      let contractsBefore;
+      ({
+        protocolDiamondAddress,
+        protocolContracts: contractsBefore,
+        mockContracts,
+      } = await deploySuite(deployer, oldVersion));
+      //      ({ twinHandler} = contractsBefore);
+
+      // Populate protocol with data
+      preUpgradeEntities = await populateProtocolContract(
+        deployer,
+        protocolDiamondAddress,
+        contractsBefore,
+        mockContracts,
+        oldVersion
+      );
+
+      // Get current protocol state, which serves as the reference
+      // We assume that this state is a true one, relying on our unit and integration tests
+      const protocolContractState = await getProtocolContractState(
+        protocolDiamondAddress,
+        contractsBefore,
+        mockContracts,
+        preUpgradeEntities,
+        oldVersion
+      );
+
+      ({ accountContractState } = protocolContractState);
+
+      // upgrade clients
+      await upgradeClients(newVersion);
+
+      // Upgrade protocol
+      ({ accountHandler } = await upgradeSuite(newVersion, protocolDiamondAddress, {
+        accountHandler: "IBosonAccountHandler",
+        orchestrationHandler: "IBosonOrchestrationHandler",
+      }));
+
+      snapshot = await ethers.provider.send("evm_snapshot", []);
+
+      // Get new account handler contract
+      const contractsAfter = {
+        ...contractsBefore,
+        accountHandler,
+      };
+
+      // This context is placed in an uncommon place due to order of test execution.
+      // Generic context needs values that are set in "before", however "before" is executed before tests, not before suites
+      // and those values are undefined if this is placed outside "before".
+      // Normally, this would be solved with mocha's --delay option, but it does not behave as expected when running with hardhat.
+      context(
+        "Generic tests",
+        getGenericContext(
+          deployer,
+          protocolDiamondAddress,
+          contractsBefore,
+          contractsAfter,
+          mockContracts,
+          protocolContractState,
+          preUpgradeEntities,
+          snapshot,
+          newVersion
+        )
+      );
+      //
+    } catch (err) {
+      // revert to latest version of scripts and contracts
+      revertState();
+      // stop execution
+      assert(false, `Before all reverts with: ${err}`);
+    }
+  });
+
+  afterEach(async function () {
+    // Revert to state right after the upgrade.
+    // This is used so the lengthly setup (deploy+upgrade) is done only once.
+    await ethers.provider.send("evm_revert", [snapshot]);
+    snapshot = await ethers.provider.send("evm_snapshot", []);
```

Comment:
We have a util function `revertToSnapshot` that can be used instead.

Reply:
done